fix: Avatar changes when from offline to online

anyongjitiger opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/28518
PROPOSAL: https://github.com/Expensify/App/issues/28518#issuecomment-1759769523


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Launch app
2. Tap on profile---Preferences
3. Toggle on forced offline
4. Tab FAB and start chat
5. Input email with default avatar and tab on create chat
6. Look at the avatar in the report
7. Tap on profile---Preferences
8. Toggle off forced offline
9. Go back to the chat, look at the avatar and make sure it will not change.
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as QA steps
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Launch app
2. Tap on profile---Preferences
3. Toggle on forced offline
4. Tab FAB and start chat
5. Input email with default avatar and tab on create chat
6. Look at the avatar in the report
7. Tap on profile---Preferences
8. Toggle off forced offline
9. Go back to the chat, look at the avatar and make sure it will not change.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/9712610/cc8cfd4b-d97e-433f-9f64-93844b5aed38
</details>





<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/9712610/33e2d682-296f-424c-8407-640203f30aa1
</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/9712610/65477cb5-a078-4fc0-a38b-f30219e4adcc


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/9712610/a970752c-7a86-4157-8bb2-389525e8a9fb


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/9712610/7879fad1-2eaa-4c5d-8c4b-1dce3e67ad4a


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/9712610/f2ea2bc8-726f-4947-b6dd-9f271d92f511


</details>

anyongjitiger committed:
fix: Avatar changes when from offline to online

anyongjitiger committed:
fix: Avatar changes when from offline to online

anyongjitiger commented:
@jjcoffee 
I'm sorry, I accidentally closed the previous pull request.
[Closed PR](https://github.com/Expensify/App/pull/31896)
 I have updated the code and sign all my commits according to your suggestions. Please review. 
Sorry again.

jjcoffee commented:
Thanks! I'll take a look tomorrow.

jjcoffee commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

jjcoffee commented:
@anyongjitiger As I mentioned on the previous PR, please link to the issue itself next to the $, rather than your proposal.

jjcoffee commented:
> There are some special cases, such as deleting a custom avatar, where the updated user ID from the backend is used to calculate the avatar URL. In such cases, it is necessary to recalculate the original user ID using the user's email.

@anyongjitiger Just tested this case that you [mentioned](https://github.com/Expensify/App/pull/31896#discussion_r1407311882) on the closed PR, but it doesn't seem to work:

https://github.com/Expensify/App/assets/27287420/02cba973-9ac9-4e89-a300-f6759697bb74

jjcoffee commented on src/libs/UserUtils.ts:
```diff
@@ -83,21 +100,39 @@ function getDefaultAvatar(accountID = -1): React.FC<SvgProps> {
-    const accountIDHashBucket = ((accountID % CONST.DEFAULT_AVATAR_COUNT) + 1) as AvatarRange;
-
+    let accountIDHashBucket: AvatarRange;
+    if (avatarURL && /images\/avatars\/default-avatar_\d+\./.test(avatarURL)) {
```
Let's also add an explanatory comment here as it won't be super clear why we're extracting the ID from `avatarURL`.

anyongjitiger replied:
Comment added

anyongjitiger committed:
Merge remote-tracking branch 'upstream/main'
merge from origin branch

anyongjitiger commented:
> > There are some special cases, such as deleting a custom avatar, where the updated user ID from the backend is used to calculate the avatar URL. In such cases, it is necessary to recalculate the original user ID using the user's email.
> 
> @anyongjitiger Just tested this case that you [mentioned](https://github.com/Expensify/App/pull/31896#discussion_r1407311882) on the closed PR, but it doesn't seem to work:
> 
>  desktop-chrome-remove-avatar-2023-11-29_15.43.20.mp4

@jjcoffee 
It seems that the backend returned a default avatar, and it used the avatar of "newDot" instead of "oldDot". 
This is the reason why the avatar changed twice after deleting it.
Do we need to change the original method when deleting avatars, instead of following what is written in the comments?

comment: `// We want to use the old dot avatar here as this affects both platforms.`

![截屏2023-11-30 09 51 56](https://github.com/Expensify/App/assets/9712610/0331f263-0629-4c96-8b55-b2f9f9563f45)

anyongjitiger committed:
Merge remote-tracking branch 'upstream/main'
merge

anyongjitiger committed:
Merge remote-tracking branch 'upstream/main'
merge

anyongjitiger commented:
@jjcoffee 
I tested the scenario in the [PR](https://github.com/Expensify/App/pull/19521#issue-1723790293), and when I set isNewDot to true, it appears to be working fine.

https://github.com/Expensify/App/assets/9712610/6df58d39-caac-446d-b991-c0f4c510cb9e

jjcoffee commented:
@anyongjitiger Great, thanks! To clarify, is that with reverting the default `isNewDot = true` back to `isNewDot = false`?

Can you also address my other comments?

anyongjitiger committed:
Merge remote-tracking branch 'upstream/main'

anyongjitiger committed:
fix: remove redundant validation

anyongjitiger committed:
fix: use newDot avatar

anyongjitiger committed:
Merge remote-tracking branch 'upstream/main'
merge

anyongjitiger commented on src/libs/UserUtils.ts:
```diff
@@ -83,21 +100,39 @@ function getDefaultAvatar(accountID = -1): React.FC<SvgProps> {
-    const accountIDHashBucket = ((accountID % CONST.DEFAULT_AVATAR_COUNT) + 1) as AvatarRange;
-
+    let accountIDHashBucket: AvatarRange;
+    if (avatarURL && /images\/avatars\/default-avatar_\d+\./.test(avatarURL)) {
```
Alright, I have removed it.

jjcoffee commented on src/libs/UserUtils.ts:
```diff
@@ -83,21 +100,47 @@ function getDefaultAvatar(accountID = -1): React.FC<SvgProps> {
+    if (allPersonalDetails?.[accountID]) {
+        if (allPersonalDetails[accountID].login) {
+            email = allPersonalDetails[accountID].login;
+        } else if (allPersonalDetails[accountID].displayName) {
```
I wonder how useful this is, since the displayName could have changed and may no longer be an email?

anyongjitiger replied:
OK, I have already updated my PR. Thank you.

jjcoffee commented on src/libs/UserUtils.ts:
```diff
@@ -83,21 +100,47 @@ function getDefaultAvatar(accountID = -1): React.FC<SvgProps> {
+        }
+    }
+    if (email) {
+        originAccountID = generateAccountID(email);
```
If I understand right, I think this would clearer if it was named `originalOptimisticAccountID`?

anyongjitiger replied:
OK, I have updated it to originalOptimisticAccountID.

anyongjitiger committed:
merge branch expensify main

anyongjitiger committed:
fix: remove param in getDefaultAvatarURL

anyongjitiger committed:
merge

anyongjitiger commented on src/libs/UserUtils.ts:
```diff
@@ -83,21 +100,47 @@ function getDefaultAvatar(accountID = -1): React.FC<SvgProps> {
     // on a simple modulo operation of their login number. Note that Avatar count starts at 1.
-    const accountIDHashBucket = ((accountID % CONST.DEFAULT_AVATAR_COUNT) + 1) as AvatarRange;
 
+    // When we create a chat with a new user, an optimistic ID is used, and the avatar is generated based on it.
```
OK, I have updated the comment.

anyongjitiger commented:
> Okay I've tested and I don't see any issues in oldDot when deleting a custom avatar (even when the avatar ID exceeds `OLD_DEFAULT_AVATAR_COUNT` (which is 8)). So I think we're good to just remove the `isNewDot` param if you agree? (Don't forget to update all instances of `getDefaultAvatarURL` to remove the extra param).
> 
> Please can you also update the linked issue in the description (after the `$`) so that it links to the issue and not your proposal? Thanks!

@jjcoffee 
I have removed the `isNewDot` param and updated the linked issue. 
Please review, thank you.

jjcoffee commented:
Thanks! Code is looking good, I will just give it a retest tomorrow and then we should be good to go.

jjcoffee reviewed (approved):
Tests well, LGTM!

jjcoffee commented:
@anyongjitiger Sorry just noticed you left a trailing `#issue` in the issue link, which is breaking automation. Can you remove that so that it's just the link to the issue?

jjcoffee commented:
@Julesssss All yours!

anyongjitiger committed:
fix: prettier

jjcoffee commented on src/libs/UserUtils.ts:
```diff
@@ -13,6 +17,12 @@ type AvatarSource = React.FC<SvgProps> | string;
+Onyx.connect({
+    key: ONYXKEYS.PERSONAL_DETAILS_LIST,
+    callback: (val) => (allPersonalDetails = _.isEmpty(val) ? {} : val),
+});
```
The link doesn't work for me for some reason :sweat_smile: Can you post some more detail here maybe?

Julesssss replied:
Bump on this question

jjcoffee replied:
Managed to get the link to load by pasting it into a chat with myself on Slack! :laughing: 

As far as I can tell we aren't memoizing any of the calls to `getDefaultAvatarURL` so I don't think the issue is particularly relevant here. Many of the calls to `getDefaultAvatarURL` are also from `ReportUtils` which also uses this pattern so it wouldn't be a very useful optimization :smile:

jjcoffee replied:
@Julesssss Friendly bump on [the above](https://github.com/Expensify/App/pull/32090#discussion_r1422384487) - I think we should be good to merge!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7256301720) to staging by https://github.com/Julesssss in version: 1.4.14-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7256301720) to staging by https://github.com/Julesssss in version: 1.4.14-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7284210064) to production by https://github.com/jasperhuangg in version: 1.4.14-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

Santhosh-Sellavel commented on src/libs/UserUtils.ts:
```diff
@@ -83,22 +100,35 @@ function getDefaultAvatar(accountID = -1): React.FC<SvgProps> {
+    // But the avatar link still corresponds to the original ID-generated link. So we extract the SVG image number from the backend's link instead of using the user ID directly
+    let accountIDHashBucket: AvatarRange;
+    if (avatarURL) {
+        const match = avatarURL.match(/(?<=default-avatar_)\d+(?=\.)/);
```
we have one more default avatar prefix  `avatar_` which is missed here, that caused a regression more details [here](https://github.com/Expensify/App/issues/33452#issuecomment-1867016966)
